src/main/java/org/cbioportal/session_service/web/SessionServiceController.java

```diff
@@ -56,55 +62,110 @@ public SessionServiceController(SessionRepository sessionRepository)
         this.sessionRepository = sessionRepository;
     }
     
-    @RequestMapping(method = RequestMethod.POST)
-    public Session addSession(@RequestBody String data) 
+    @RequestMapping(method = RequestMethod.POST, value="/{source}/{type}")
+    public Map<String, String> addSession(@PathVariable String source, 
+        @PathVariable String type, 
+        @RequestBody String data) 
+    { 
+        Session session = new Session(source, type, data); 
+        try {
+            sessionRepository.saveSession(session); 
+        } catch (DuplicateKeyException e) {
+            // find session and return it
+            // need the JSON object data, not the string passed
+            session = sessionRepository.findOneBySourceAndTypeAndData(source, 
+                type, 
+                session.getData()); 
+        } catch (ConstraintViolationException e) {
+            throw new SessionInvalidException(buildConstraintViolationExceptionMessage(e));
+        }
+        Map<String, String> map = new HashMap<String, String>();
+        map.put("id", session.getId());
+        return map;
+    }
+
+    @RequestMapping(method = RequestMethod.GET, value="/{source}/{type}")
+    public Iterable<Session> getSessions(@PathVariable String source, 
+        @PathVariable String type)
     {
-        return sessionRepository.save(new Session(data)); 
+        return sessionRepository.findBySourceAndType(source, type);
     }
 
-    @RequestMapping(method = RequestMethod.GET)
-    public Iterable<Session> getSessions()
+    @RequestMapping(method = RequestMethod.GET, value="/{source}/{type}/query")
+    public Iterable<Session> getSessionsByQuery(@PathVariable String source, 
+        @PathVariable String type, 
+        @RequestParam(name="field") String field,
+        @RequestParam(name="value") String value)
     {
-        return sessionRepository.findAll();
+        List<Session> sessions = sessionRepository.findBySourceAndTypeAndQuery(source, type, field, value);
+        if (sessions.size() != 0) {
+            return sessions;
+        }
+        throw new SessionNotFoundException("field=" + field + "&value=" + value);
```

Comment:
Take a look at the use of @ExceptionHandler here:

https://github.com/ersinciftci/mutation/blob/master/web/src/main/java/org/cbioportal/mutation/web/rest/MutationController.java